File: Scala/algorithmic_toolbox/src/main/scala-2.12/greedy/MaximizingRevenue.scala
package greedy

object MaximizingRevenue {
  private def convertToIntList(line: String): List[Int] = line.split(" ").map(_.toInt).toList

  def calcMaximalRevenue(profitPerClick: List[Int], averageClickPerDay: List[Int]): Long =
    profitPerClick.sorted.zip(averageClickPerDay.sorted).foldLeft(0L){ case (acc, (p, a)) => acc + p * a.toLong }

  def main(args: Array[String]): Unit = {
    val reader: Iterator[String] = scala.io.Source.stdin.getLines()
    val _ :Int = reader.next().toInt
    val profitPerClick: List[Int] = convertToIntList(reader.next())
    val averageClickPerDay: List[Int] = convertToIntList(reader.next())
    val result: Long = calcMaximalRevenue(profitPerClick, averageClickPerDay)
    println(result)
  }
}

File: OCaml/algorithmic_toolbox/greedy/maximizing_revenue.ml
let convert_to_intlist line = List.map int_of_string Str.(line |> split (regexp " "))


let calc_maximal_revenue profit_per_click average_click_per_day =
    List.fold_left2
        (fun acc p c -> acc + p * c) 0 (List.sort compare profit_per_click) (List.sort compare average_click_per_day)


let () =
    let _ = read_int() in
    let profit_per_click = convert_to_intlist (read_line()) in
    let average_click_per_day = convert_to_intlist (read_line()) in
    print_int (calc_maximal_revenue profit_per_click average_click_per_day);
    print_newline()
